src/rpc/validation.rs

```diff
@@ -74,7 +69,8 @@ where
                 );
             })
     }
-    async fn validate_inner(&self) -> RpcResult<()> {
```

Comment:
So if I understand correctly the problem arose in this function, because it was an `async` function with no intermediate `await` calls, making it "block" the thread and preventing it from yielding and switching to other tasks. Therefore once we would have reached a number of requests in progress equal to the number of available "core threads" (which by default seems to be equal to number of available CPU core on the host), new requests would have ended up in a queue. Running this code in a blocking thread solves that and means that we would only reach this situation when the much higher limit of threads in the block pool is reached.

It'd seem that an alternative approach would have been to declare the individual steps (`check_gas_limit` etc.) async and thereby allow the thread to yield on each substep, but that would not have helped that much because the majority of the compute comes from 2 operations (`execute_and_verify_block` and `validate_state_root`).

@mattsse Is that thinking / understanding correct ? 
Note that my understanding of this issue is based mostly on [this article](https://ryhl.io/blog/async-what-is-blocking/) as well as the [tokio documentation](https://docs.rs/tokio/latest/tokio/index.html#cpu-bound-tasks-and-blocking-code)

Reply:
the table in the article highlights this as well.

> Running this code in a blocking thread solves that and means that we would only reach this situation when the much higher limit of threads in the block pool is reached.

the block pool scales and by default has a very high limit of available slots.

but the issue that long running tasks can negatively impact the overall tokio's executor performance, hence those should be moved to a blocking task, or rayon

Reply:
Makes sense, thanks.